### DISCUSSION TITLE
Using qiBullet with ROS: odometry broadcasting problems

### DISCUSSION THREAD

**Discussion by ysys98** [softbankrobotics-research/qibullet#78] [Q&A] [answered]
#76 
Still this problem,  I found that it should be the problem of parameter of  _ broadcastodometry (self, odometry_publisher), so I passed odom_pub into parameter to Is wrong?
<img width="620" alt="3" src="https://user-images.githubusercontent.com/98088332/159479657-5a063852-f8b9-49a0-96df-bbd1fd226bfb.png">

And I changed odometry_ publisher file changed from .CPP to .Py file
<img width="565" alt="1" src="https://user-images.githubusercontent.com/98088332/159479131-daee2f08-5a7f-45ae-a72b-ebedf659e4a1.png">
<img width="645" alt="2" src="https://user-images.githubusercontent.com/98088332/159479171-16fca7d9-38c7-4f98-89b4-d99e929c4887.png">

And then when I run 
It shows the error log:
<img width="433" alt="4" src="https://user-images.githubusercontent.com/98088332/159480676-91055988-1e32-46e9-a6d1-e7a67d5582be.png">

"odom_pub" is not defined

**Comment by mbusy** [comment]
I don't really understand what you're trying to do here, but as specified in the comments of the method, `_broadcastOdometry` is an internal method and shouldn't be used in a script. The method is defined in the `RosWrapper` class, and the ROS wrapper for each robot (PepperRosWrapper etc) inherit from that class. You don't need to call the `_broadcastOdometry` method manually, it's done automatically when you launch the wrapper. 

Here's a quick example:
```python
if __name__ == "__main__":
    simulation_manager = SimulationManager()
    wrap = PepperRosWrapper()
    camera_id = PepperVirtual.ID_CAMERA_BOTTOM
    pepper = simulation_manager.spawnPepper(client, spawn_ground_plane=True)

    wrap.launchWrapper(pepper, "/naoqi_driver")

    handle = pepper.subscribeCamera(camera_id)

    try:
        rospy.spin()

    except KeyboardInterrupt:
        pass
    except rospy.ROSInterruptException:
        pass
    finally:
        wrap.stopWrapper()
        simulation_manager.stopSimulation(client)
```

The `launchWrapper` is all you need, the odometry will be broadcasted

**Comment by mbusy** [comment]
I'm going to move this issue into the discussions and widen the scope of the issue a bit, since I think the issue has more to do with the joint usage of ROS with qiBullet. If a specific issue with the simulation arises, we'll create a more specific issue

**Reply by ysys98** [reply]
> 我将把这个问题移到讨论中并稍微扩大问题的范围，因为我认为这个问题更多地与 ROS 与 qiBullet 的联合使用有关。如果模拟出现特定问题，我们将创建一个更具体的问题
@mbusy I'm sorry I didn't express the question clearly. Thank you for your patience. As you said, I used the  launchwrapper  method after getting your reply, but the odometry hasn't been broadcast yet,I still can't find odom in the Fixed Frame of rviz
Here is the code of your example I used：

`#!/usr/bin/env python
#coding: utf-8
import os
import sys
import rospy
import pybullet
import pybullet_data
from qibullet import PepperVirtual
from qibullet import PepperRosWrapper
from qibullet import SimulationManager

if __name__ == "__main__":
    simulation_manager = SimulationManager()
    client = simulation_manager.launchSimulation(gui=True)
    wrap = PepperRosWrapper()
    camera_id = PepperVirtual.ID_CAMERA_BOTTOM
    pepper = simulation_manager.spawnPepper(client, spawn_ground_plane=True)

    wrap.launchWrapper(pepper, "/naoqi_driver")

    handle = pepper.subscribeCamera(camera_id)
   
    try:
        rospy.spin()

    except KeyboardInterrupt:
        pass
    except rospy.ROSInterruptException:
        pass
    finally:
        wrap.stopWrapper()
        simulation_manager.stopSimulation(client)`
<img width="977" alt="555" src="https://user-images.githubusercontent.com/98088332/159837591-91cdc512-783d-412e-b0cd-6443d97657e4.png">
And when I run, he always appears such a log
**
NODES

ROS_MASTER_URI=http://localhost:11311
process[robot_state_publisher_roskinetic_VirtualBox_8752_6197811648838051663-1]: started with pid [8901]
Stopping the ROS wrapper: function takes at most 2 arguments (3 given)
**

**Reply by mbusy** [reply]
Ah alright, that's indeed more problematic. This message occurs when an exception is caught in the _spin fonction of the ROS wrapper. I'll try to recreate that bug, using the code you provided. I have a couple more questions:

- What OS do you use ? (you seem to be using virtualbox, what is the OS you're using in there, and what is the default OS on your computer?)
- You seem to be using ROS kinetic?
- Did you install qibullet from source or via pip?
- What is your version of qibullet? (You can get it through python, import qibullet and display `qibullet.__version__`)
- You're only using the code you provided and RViz, nothing else that could interfere?

**Reply by ysys98** [reply]
> Ah alright, that's indeed more problematic. This message occurs when an exception is caught in the _spin fonction of the ROS wrapper. I'll try to recreate that bug, using the code you provided. I have a couple more questions:
> 
> * What OS do you use ? (you seem to be using virtualbox, what is the OS you're using in there, and what is the default OS on your computer?)
> * You seem to be using ROS kinetic?
> * Did you install qibullet from source or via pip?
> * What is your version of qibullet? (You can get it through python, import qibullet and display `qibullet.__version__`)
> * You're only using the code you provided and RViz, nothing else that could interfere?

@mbusy Thank you very much for your reply. 
The default system of my computer is windows, and I also use the virtual machine of VirtualBox
I use VirtualBox, Ubuntu 16 04 and ROS kinetic，and I install qibullet by pip，
my qibullet version is 1.4.4
<img width="409" alt="666" src="https://user-images.githubusercontent.com/98088332/159974404-5072c938-7a58-4caa-b225-1a974c66e276.png">

I just only using the code you provided and launch the RViz
here is my rviz launch：
`<launch>
    <node pkg="rviz" type="rviz" name="rviz" />
    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" output="screen" />
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen" />
    <node pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" name="joint_state_publisher_gui" output="screen" />
</launch>`

**Reply by mbusy** [reply]
Thanks for the info. Alright, you don't need to launch the state publishers from your launchfile, normally the ROS wrapper also takes care of that. The only requirement of the ROS wrapper is that a roscore already exists when `launchWrapper` is called from Python. You can try to remove the `joint_state_publisher` and `robot_state_publisher` from your launchfile.

But I'm not sure that this is the root of the problem though

**Reply by ysys98** [reply]
> Thanks for the info. Alright, you don't need to launch the state publishers from your launchfile, normally the ROS wrapper also takes care of that. The only requirement of the ROS wrapper is that a roscore already exists when `launchWrapper` is called from Python. You can try to remove the `joint_state_publisher` and `robot_state_publisher` from your launchfile.
> 
> But I'm not sure that this is the root of the problem though

@mbusy Thank you for your reply. I know I must start roscore first every time. I have deleted the joint_state_publisher and robot_state_publisher from launchfile. But it still can't show odom at RVIZ Fixed Frame
<img width="623" alt="999" src="https://user-images.githubusercontent.com/98088332/159976907-39ba894d-05ee-45bd-97e2-3962cc2b5ac9.png">
Every time I run, it will appear such an log：
NODES

ROS_MASTER_URI=http://localhost:11311
process[robot_state_publisher_roskinetic_VirtualBox_2378_6291095413713276270-1]: started with pid [2582]
Stopping the ROS wrapper: function takes at most 2 arguments (3 given)

Could it be the problem here？

**Reply by mbusy** [reply]
> Could it be the problem here？

The problem is definitely there, but first I wanted to make sure that nothing was interfering with the information published by the wrapper. Now it should be the case, an now in RViz we don't have any states received (as you can see the meshes of the robot are all merged together, the frames are not correctly placed). That is quite weird, I can't reproduce your problem on my side...

> Stopping the ROS wrapper: function takes at most 2 arguments (3 given)

I'm going to redo the tests using kinetic instead of melodic, but theoretically it shouldn't make much of a difference

**Reply by ysys98** [reply]
@mbusy Thank you for your patience in answering my questions
Is there a possible pybullet module problem? When I first got your reply 10 days  ago, I found that my import pybullet was not prompted in vscode. If vscode can find this module, it will display green, but at that time, the pybullet was white. I installed it using pip, so I uninstalled it and found this problem in the pybullet of GitHub, at https://github.com/bulletphysics/bullet3/issues/4071
and then I installed it using this method, Then vscode shows that the import pybullet is green instead of white, but when I open ros_wrapper.py
here is a line：        [vx, vy, vz], [wx, wy, wz] = pybullet.getBaseVelocity(
Vscode also shows that it was not found getBaseVelocity，
getBaseVelocity is white. When you click it, there is no prompt. I found that other pybullet methods, vscode show them are also white without any prompt，
this is a little strange. Qibullet methods or functions will display some prompts in vscode, and they are not white. Is it my pybullet download error？

**Reply by ysys98** [reply]
@mbusy By the way, when I first used qibullet ten days ago, I used the example in the official tutorial robot_ ros_ test. py
The first time I ran it, he appeared in such a log:
Stopping the ROS wrapper: function takes at most 2 arguments (3 given)

My reply above may not be clear. I intercepted some pictures, which may be more intuitive
![5](https://user-images.githubusercontent.com/98088332/159986933-9f701500-1e83-4d34-b995-10b4f3c6597b.jpg)
![4](https://user-images.githubusercontent.com/98088332/159986946-a885ef4f-b233-439d-bd2e-5b5998e1ced3.jpg)
![3](https://user-images.githubusercontent.com/98088332/159986957-b1edd42f-6769-40ff-b4fd-4b17e3e48110.jpg)
![2](https://user-images.githubusercontent.com/98088332/159986962-49b5959e-56a9-46b5-85e7-48d560290ddf.jpg)
![1](https://user-images.githubusercontent.com/98088332/159986965-9a3ee7f1-dd08-4a8a-84a0-60b89ca64bb6.jpg)

Vscode is not found pybullet‘s method , but it doesn't affect the operation of the program, does it?
Or did I not downloaded pybullet correctly ？
I don't know if this is the reason why I made a mistake. If not, please forgive me

**Reply by mbusy** [reply]
Your comments are clear don't worry 😉 I found the source of the problem, it's linked to a recent update of pybullet, that broke the compatibility in `getBaseVelocity`. The method still exists, but different parameters are expected, it's an easy fix though! removing the line 207 of ros_wrapper.py should do the trick. I'll correct that, test it and publish an updated qibullet package

**Comment by mbusy** [comment, marked as answer]
Alright @ysys98, the hotfix/ros_odometry branch should take care of the problem. Could you try to uninstall qibullet and reinstall it from source using this branch? (the [wiki](https://github.com/softbankrobotics-research/qibullet/wiki/Tutorials:-Installation#from-source) should help you with that, checkout the `hotfix/ros_odometry` branch after cloning the repo, and before installing) It works for me, if it works for you I'll package an official release

**Reply by ysys98** [reply]
@mbusy sorry, where can I find the hotfix/ros_odometry branch？In which？
Should I do this？
step 1.uninstall qibullet
step 2.How do I reinstall it from source using this branch? where is repo？Is git clone https://github.com/softbankrobotics-research/qibullet.git?
step 3.how do I checkout the hotfix/ros_odometry branch after cloning the repo？
I'm really sorry, I don't know enough about step 2 and step 3

**Reply by mbusy** [reply]
Sure, in the order you specified
```bash
pip uninstall qibullet
git clone https://github.com/softbankrobotics-research/qibullet.git
cd qibullet
git checkout hotfix/ros_odometry
python setup.py install --user
```

**Reply by ysys98** [reply]
@mbusy  Thank you very much，it wroked！
but whether pepper or Nao， half of them body is still under the ground when I choice odom at Fixed Frame
<img width="436" alt="2000" src="https://user-images.githubusercontent.com/98088332/159998930-ac926f3f-4919-4696-875c-25717964f710.png">
![2001](https://user-images.githubusercontent.com/98088332/159999263-f0757a6d-c2e4-4e0d-940d-f06f0b009f5f.jpg)

**Reply by mbusy** [reply]
Right, that's another issue depending on if you use the robot model defined in the classical naoqi driver or the softbankrobotics-research version of it. I guess the wrapper could use a dynamic solution to deal with that, depending on the model that is used

**Reply by ysys98** [reply]
> Right, that's another issue depending on if you use the robot model defined in the classical naoqi driver or the softbankrobotics-research version of it. I guess the wrapper could use a dynamic solution to deal with that, depending on the model that is used

@mbusy How can I solve this issue，I mean, I don't quite understand about the classical naoqi driver or the softbankrobotics-research version，I don't know which robot model I'm using

**Reply by mbusy** [reply]
The classical one I guess, this will require a bit more work thank the quick fix. I'll let you know if I have time to add something to the current hotfix

**Reply by ysys98** [reply]
> The classical one I guess, this will require a bit more work thank the quick fix. I'll let you know if I have time to add something to the current hotfix

@mbusy  Thank you for your help and patience, and I look forward to your early reply !

**Reply by mbusy** [reply]
@ysys98 the last commit on the hotfix branch should help. The base_footprint frame of Pepper is still a bit below ground when the robot is referenced in odom though. I believe that it's because different Pepper models are used in qibullet and the ROS stack. Fixing that would require to update one stack, or consider that the base_link of Pepper in the official ROS stack is it's footprint instead of its torso. either way, it's probably not something that will be done quicky. I still have an update of the driver to publish when I have the time, I'll try to add that to the update

**Reply by ysys98** [reply]
@mbusy Thank you very much！ The NAO is also due to the same problem, right?
 I will always wait for your update! ！！！！！！！
thank you

**Reply by ysys98** [reply]
@mbusy Thank you a lot！It worked！
<img width="416" alt="1" src="https://user-images.githubusercontent.com/98088332/160222639-f6664781-eb86-40fb-9245-a46df878708a.png">
Because there are only Nao machines in our school laboratory, so I can simulate next

**Reply by ysys98** [reply]
@mbusy May I ask you something，how I can realize the automatic navigation of Nao. Some algorithms such as gmapping in ROS are suitable for wheeled robots such as pepper, but not for biped robots such as Nao
Can the method mentioned in #47  realize autonomous navigation？

**Reply by mbusy** [reply]
The methods mentioned in this issue will give you a solution to generate walking motion for the NAO robot in qiBullet. On top of that, using gmapping for instance will allow you to realize autonomous navigation.

As mentionned in the issue, we don't plan on adding a walk controller to the project, but we could consider merging a PR that would add such a feature

**Reply by ysys98** [reply]
@mbusy Which one should I use? [https://github.com/asherikov/oru_walk_module](url) or [https://github.com/mrsp/nao_walk](url)?

**Reply by mbusy** [reply]
I didn't use any of them, so I wouldn't know which one is better. Also, both seem to rely on the naoqi to work (most probably to read joint angles and sensor values, and control joints), so some tweaking would be required to interface them with qiBullet.

**Reply by ysys98** [reply]
@mbusy Thank you very much！I will learn it first. The ultimate goal is to try to use him in a real Nao robot. Anyway, I should try to learn this controller first

**Reply by mbusy** [reply]
Sure. The 1.4.5 version has been uploaded, fixing the original odom broadcasting bug

**Reply by ysys98** [reply]
@mbusy 
Thanks! Thanks for your help!
